embezzlement (#15221)
PostHog/posthog.com: contents/blog/optimizing-agent-cost.mdx

Reviewer: I think choosing present tense can make this a bit of an awkward read in places, but in preference to shipping faster, just keeping as is mostly.

---
title: "How we caught our token-embezzling Wizard"
date: 2026-01-08
author:
  - vincent-ge
# featuredImage: >-

> #     ???

Reviewer: We can reuse this one, we can change the text if you want. 
```suggestion
featuredImage: >-
  https://res.cloudinary.com/dmukukwp6/image/upload/Film_Noir_Hog_Evals_beta_announcement_blog_95685493eb.png

```

For initial prompt size `N` and average input size `K` over `T` turns, you can estimate the growing context window with this approximation:

```
N + (N+K) + (N+2K) + ... + (N+TK) ≈ T*N + K*T²/2
```

That's a lot of tokens!

If the Wizard runs out of context, which happens nearly every run, the context gets compacted.

Remember that compactions [aren't free](https://platform.claude.com/docs/en/build-with-claude/compaction#understanding-usage), I'll be paying for:
- The cache reads on the existing context window
- The output of that compaction API call
- And post compaction, the cost of cache writes of the new compacted context

### What if I break up the query?

Since the Wizard tracks every event captured in `posthog-events.json`, the `conclude` step doesn't need any prior context to do its job. Why's the Wizard carrying all that context around? 

If I can lower costs by clearing this extra context, I can prove the Wizard's wasteful spending.

> The easiest way to clear the conversation is just to start a new `query()` or conversation. I made the Wizard switch to running a new query for every step of the flow, and using `continue_conversation` to pass context between steps `begin` through `revise`, where prior context is relevant. The final query call for step `conclude` will run on a fresh context. 

Reviewer: ```suggestion
The easiest way to clear the conversation is just to start a new `query()` or conversation. I made the Wizard switch to running a new query for every step of the flow, and using `continue_conversation` to pass context between steps `begin` through `revise`, where prior context is relevant. The final query call for step `conclude` runs on a fresh context. 
```

    end

    subgraph Q3["query() #3 — Conclude"]
        F["1.3 Conclude<br>MCP & Report"]
    end

    Q1 --> Q2 --> Q3

    style Q1 fill:#e3f2fd,stroke:#2c5f8a
    style Q2 fill:#e8f5e9,stroke:#5a9a18
    style Q3 fill:#fce4ec,stroke:#9a0114
    style A fill:#4a90d9,stroke:#2c5f8a,color:#fff
    style C fill:#f5a623,stroke:#c47d0e,color:#fff
    style D fill:#7ed321,stroke:#5a9a18,color:#fff
    style E fill:#f5a623,stroke:#c47d0e,color:#fff
    style F fill:#d0021b,stroke:#9a0114,color:#fff
```

This was more promising. I saw ideal cases result in a 70% cost reduction.

But it wasn't a clear winner. Some runs were still more expensive than running everything in a single agent loop. The cost difference between the cheapest and most expensive runs is $8 USD, there is a huge variation between the cheapest and most expensive runs.

| Metric | Mean | Std | Min | Max |
|---|---|---|---|---|
| numTurns | 82.4 | 36.5 | 20 | 142 |
| totalCostUsd | 6.29 | 2.48 | 2.06 | 10.83 |
| durationMs | 530,795 | 210,259 | 254,488 | 1,523,166 |
| inputTokens | 5.95M | 4.39M | 371k | 12.28M |

> As counter-intuitive as it sounds, the Wizard running eveerything in one giant loop and carrying around all that context is actually very efficient.

Reviewer: ```suggestion
As counter-intuitive as it sounds, the Wizard running everything in one giant loop and carrying around all that context is actually very efficient.
```

### What about subagents?

> Another common way to clear the context is to use [subagent](https://platform.claude.com/docs/en/agent-sdk/subagents). I tried instructing the main agent to run step `conclude`'s instructions in its own subagent. This approach also has a few issues.

Reviewer: ```suggestion
Another common way to clear the context is to use [subagent](https://platform.claude.com/docs/en/agent-sdk/subagents). I tried instructing the main agent to run the `conclude` step instructions in its own subagent. This approach also has a few issues.
```

The first issue is that even though this subagent sits under the same `query()` call, it doesn't share the parent `query()` call's cache. I face the same issue of incurring extra cache write costs.

The second, much bigger problem, is that subagents don't seem to reliably terminate. There are many [open issues](https://github.com/anthropics/claude-code/issues/19926) on this topic. This causes one of two things to happen:
The Wizard hangs and never terminates.
The Wizard sets timeouts on subagents. Sometimes this timeout is too low, and it tries again with a longer timeout. This often loops many times.

Neither cases are acceptable.

This is a dead end. Even if sending all the extra context feels wasteful, the data doesn't support it. I still don't have a case.

## Connecting the dots

While I searched through piles of Wizard logs, I noticed something strange.

There is a 150% difference between the Wizard runs with the most and least turns. Maybe I've been misled all along. What if it's not accumulated context. What if it's not just the `conclude` step. What if runaway tool calls are blowing up the number of turns, and in turn our cost?

> When I plot all the Wizard runs using multiple queries (green) against single-query agents (red), I find them to form two clusters:

Reviewer: What does phased vs unphased mean?

## The full confession

When I followed the Wizard's path on high-turn, high-cost outlier runs on the right hand cluster, everything started clicking. When I carefully interrogated the Wizard over these runs turn-by-turn, I finally had a full confession.

I can break the Wizard's token laundering schemes into three categories:

### Wizard bureaucracy

In step `conclude`, the Wizard has all it needs to know in `posthog-events.json`. However, I found that the Wizard will sometimes re-read through the code base to confirm that `posthog-events.json` is accurate.

While I appreciate that the Wizard claim that "this prevents hallucinations and ensures correctness", all these extra reads, cache writes, and compactions often double the cost of the `conclude` step.

### Compaction amnesia

Compaction amnesia occurs most commonly in larger projects. I caught the Wizard re-reading a ton of files after a big compaction.

Remember that compactions already carry a cost. Re-reading the project files causes a ton of new cache to be written, which are 12x more expensive than cache reads. 

> ### Fly-away subagents

Reviewer: ```suggestion
### 3. Fly-away subagents
```

This is the biggest culprit to exploding costs. Sometimes the Wizard uses subcontractors (subagents) to do exploration, run linters, and install packages.

At every level, the subagents edge out more inference costs. Remember that subagents don't share the parent agent's cache. This incurs expensive cache writes.

Agents can also hang and not return after completion. Sometimes the parent agent can set unreliable timeouts to await long-running subagents, often backing off exponentially with 1, 3, 5, 10 minute timeouts. Anthropic's cache has a 5 minute timeout. This means waiting more than 10 minutes for a subagent to return causes the entire cache to be rebuilt.

In an extreme case, I caught our Wizard spawning 5 sub agents, idling for 5 minutes waiting for the subagents to return, then running `KillShell` because 2 of 5 subagents didn't return. This entire process was a waste and took about 15 minutes.

When the Wizard finally returns, it runs `pnpm install` and makes all the necessary edits in just 3 minutes.

This is corruption and nepotism under the guise of "efficiency" and "parallelism".

## Justice for stakeholders

> The Wizard has been found guilty on all counts. Here's the legislation and processes we're putting in place so this never happens again.

Reviewer: This needs to be a slightly different framing if you are going to make suggestions. 

I think it should be just what we learned (and added some comments below to that end), but if you wanted to you could expand it with something like this:

```suggestion
## Justice for stakeholders

The Wizard has been found guilty on all counts. Here's the legislation and processes we're putting in place so this never happens again (and what you can do to prevent it in the first place). 
```

### Efficient caching

A bigger context window doesn't always mean higher cost. Cache writes cost 12x more than cache reads. Proposals to break up the agent's `query()` require a 12x reduction in token reads to justify the new token writes to cache.

### Reduced bureaucracy

Token costs increase quadratically as turns accumulate. Agents often waste turns under the guise of "ensuring correctness". The real issue here is that the agent doesn't know which parts of the existing context is up to date and which is not.

Designate trusted sources for agents to store important information it will reuse, and _explicitly_ prompt them to use and trust that information, like our `posthog_events.json`. This greatly reduces repeated file read tool calls.

### The subagent ban

Nepotism and token corruption go hand-in-hand. Subagents don't share a cache prefix and don't terminate reliably as of today. Many sources claim that subagents can increase efficiency by parallelizing tasks, but the current implementation is just not reliable enough.

It's often cheaper and faster to just have your one agent do it all.

### Standardizing deterministic flows

> Where possible, keep the flow deterministic. In the `conclude` step, the task has a deterministic flow. Read `posthog-events.json`, create 5 query MCP calls, 5 insight MCP calls, and add them to a dashboard with another MCP call. It can be done effectively in just 10 turns, so it should always take just 10 turns. 

Reviewer: ```suggestion
Where possible, we keep the flow deterministic. In the `conclude` step, the task has a deterministic flow. Read `posthog-events.json`, create 5 query MCP calls, 5 insight MCP calls, and add them to a dashboard with another MCP call. It can be done effectively in just 10 turns, so it should always take just 10 turns. 
```
